File: KengenWaterApp/KengenWaterApp/app/src/main/java/com/water/alkaline/kengen/ui/adapter/ChannelAdapter.java
package com.water.alkaline.kengen.ui.adapter;


import android.app.Activity;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.gms.ads.AdLoader;
import com.google.gms.ads.MyApp;
import com.google.gms.ads.databinding.LayoutAdUniversalBinding;
import com.preference.PowerPreference;
import com.water.alkaline.kengen.MyApplication;
import com.google.gms.ads.AdUtils;
import com.water.alkaline.kengen.databinding.ItemImageBinding;
import com.water.alkaline.kengen.model.main.Banner;
import com.water.alkaline.kengen.model.main.Channel;
import com.water.alkaline.kengen.ui.listener.OnChannelListener;
import com.water.alkaline.kengen.utils.Constant;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ChannelAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    Activity activity;
    List<Channel> arrayList = new ArrayList<>();
    OnChannelListener listener;

    public ChannelAdapter(Activity activity, List<Channel> arrayList, OnChannelListener listener) {
        this.activity = activity;
        this.arrayList = arrayList;
        this.listener = listener;
        setAds(false);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ItemImageBinding binding;

        public ViewHolder(ItemImageBinding itemView) {
            super(itemView.getRoot());
            binding = itemView;
        }
    }


    public void setAds(boolean isAds) {
        int PARTICLE_AD_DISPLAY_COUNT = MyApp.getAdModel().getAdsListViewCount();

        if (PARTICLE_AD_DISPLAY_COUNT > 0 && MyApp.getAdModel().getAdsOnOff().equalsIgnoreCase("Yes")) {

            arrayList.removeAll(Collections.singleton(null));
            ArrayList<Channel> tempArr = new ArrayList<>();

            if (!arrayList.isEmpty()) {
                tempArr.add(null);
            }

            for (int i = 0; i < arrayList.size(); i++) {
                if (arrayList.size() > PARTICLE_AD_DISPLAY_COUNT) {
                    if (i != 0 && i % PARTICLE_AD_DISPLAY_COUNT == 0) {
                        tempArr.add(null);
                    }
                    tempArr.add(arrayList.get(i));
                } else {
                    tempArr.add(arrayList.get(i));
                }
            }

            this.arrayList = tempArr;
        }

        if (isAds)
            notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        return arrayList.get(position) == null ? Constant.AD_TYPE : Constant.STORE_TYPE;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == Constant.AD_TYPE)
            return new AdHolder(LayoutAdUniversalBinding.inflate(LayoutInflater.from(activity), parent, false));
        else
            return new ViewHolder(ItemImageBinding.inflate(LayoutInflater.from(activity), parent, false));
    }

    public void refreshAdapter(List<Channel> arrayList) {
        this.arrayList = arrayList;
        setAds(true);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof AdHolder) {
            AdHolder adHolder = (AdHolder) holder;
            if (adHolder.binding.flAd.getChildCount() <= 0) {
                AdLoader.getInstance().showNativeList(activity, adHolder.binding);
            }
        } else {
            ViewHolder viewHolder = (ViewHolder) holder;
            Glide.with(activity).load(arrayList.get(position).getUrl())
                    .placeholder(MyApplication.getPlaceHolder())
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(viewHolder.binding.imgVideo);
            viewHolder.binding.txtVideoTitle.setText(arrayList.get(position).getName());
            viewHolder.binding.txtVideoTitle.setSelected(true);
            viewHolder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    listener.onItemClick(holder.getAdapterPosition(), arrayList.get(holder.getAdapterPosition()));
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return arrayList.size();
    }
}

File: KengenWaterApp/KengenWaterApp/app/src/main/java/com/water/alkaline/kengen/ui/adapter/ChannelAdapter.kt
package com.water.alkaline.kengen.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.google.gms.ads.AdLoader
import com.google.gms.ads.MyApp
import com.google.gms.ads.databinding.LayoutAdUniversalBinding
import com.water.alkaline.kengen.MyApplication
import com.water.alkaline.kengen.databinding.ItemImageBinding
import com.water.alkaline.kengen.model.main.Banner
import com.water.alkaline.kengen.model.main.Channel
import com.water.alkaline.kengen.ui.listener.OnChannelListener
import com.water.alkaline.kengen.utils.Constant


class ChannelAdapter(
    var activity: Context,
    var arrayList: MutableList<Channel>,
    var listener: OnChannelListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    init {
        setAds(false)
    }

    inner class ViewHolder(var binding: ItemImageBinding) : RecyclerView.ViewHolder(
        binding.root
    )


    private fun setAds(isAds: Boolean) {

        val PARTICLE_AD_DISPLAY_COUNT = MyApp.getAdModel().adsListViewCount
        if (PARTICLE_AD_DISPLAY_COUNT > 0 && MyApp.getAdModel().adsOnOff.equals(
                "Yes",
                ignoreCase = true
            )
        ) {
            arrayList.removeAll {
                it.id.equals(Constant.defaultId)
            }
            val tempArr = mutableListOf<Channel>()
            if (arrayList.isNotEmpty()) {
                tempArr.add(Channel().apply {
                    id = Constant.defaultId
                })
            }

            for (i in arrayList.indices) {
                if (arrayList.size > PARTICLE_AD_DISPLAY_COUNT) {
                    if (i != 0 && i % PARTICLE_AD_DISPLAY_COUNT == 0) {
                        tempArr.add(Channel().apply {
                            id = Constant.defaultId
                        })
                    }
                    tempArr.add(arrayList[i])
                } else {
                    tempArr.add(arrayList[i])
                }
            }

            this.arrayList = tempArr
        }

        if (isAds) notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        return if (arrayList[position].id.equals(Constant.defaultId)) Constant.AD_TYPE
        else Constant.STORE_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == Constant.AD_TYPE) AdHolder(
            LayoutAdUniversalBinding.inflate(
                LayoutInflater.from(
                    activity
                ), parent, false
            )
        )
        else ViewHolder(
            ItemImageBinding.inflate(
                LayoutInflater.from(
                    activity
                ), parent, false
            )
        )
    }

    fun refreshAdapter(arrayList: MutableList<Channel>) {
        this.arrayList = arrayList
        setAds(true)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is AdHolder) {
            val adHolder = holder
            if (adHolder.binding.flAd.childCount <= 0) {
                AdLoader.getInstance().showNativeList(activity, adHolder.binding)
            }
        } else {
            val viewHolder = holder as ViewHolder
            Glide.with(activity).load(arrayList[position]!!.url)
                .placeholder(MyApplication.getPlaceHolder())
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(viewHolder.binding.imgVideo)
            viewHolder.binding.txtVideoTitle.text = arrayList[position]!!.name
            viewHolder.binding.txtVideoTitle.isSelected = true
            viewHolder.itemView.setOnClickListener {
                listener.onItemClick(
                    holder.bindingAdapterPosition,
                    arrayList[holder.bindingAdapterPosition]
                )
            }
        }
    }

    override fun getItemCount(): Int {
        return arrayList.size
    }
}


